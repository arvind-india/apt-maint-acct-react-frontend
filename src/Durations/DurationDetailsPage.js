import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import {
    Form,
    Button,
    FormText,
    Input,
    Label,
} from 'reactstrap'

import { durationActions as actions, alertActions } from '../_actions'
import { FlashMessage } from '../_components'

let module = 'durations' // module name

export class DurationDetails extends React.Component {

  constructor(props) {
    super(props)
    const { location } = props
    let model = location.state.model // model supplied from list page
    let initializeModel = {
      id: model.id,
      key: model.key,
      value: model.value,
      effective_from: model.effective_from,
      effective_to: model.effective_to,
      remarks: model.remarks
    }
    this.state = {
      model: initializeModel, // model to edit
      submitted: false,
      adding: model.id === 0
    }
    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  canBeSaved() { // check for changes in model, if changes present, it can save
    const { model } = this.state
    if (!model.key) return false
    if (!model.value) return false
    if (!model.effective_from) return false
    if (!model.effective_to) return false
    return true // can save changes
  }

  render() {
    const { alert } = this.props
    return (
      <div>
        <h2>Duration Details</h2>
        {alert.message && <FlashMessage text={alert.message} color={alert.color} delay={2100}/>}
        {this.validateForm()}
        {this.show()}
      </div>
    )
  }
  validateForm() {
    if ( this.changedProps().length === 0 ) {
      this.validationMsg = 'No changes to save'
      this.formValid = false
      return null
    }
    if( this.canBeSaved() ) {
      this.validationMsg = 'Save Changes'
      this.formValid = true
      return null
    }
    // finally, if reached here
    this.validationMsg = 'Missing "Required Data"...'
    this.formValid = false
  }

  show(){
    const { adding } = this.state
    const { authzn } = this.props
    let title = adding?'Add':authzn.allowsEdit?'Edit':'View'
    return <Form id="durationDetailsForm" onSubmit={this.handleSubmit} className="grid-form">
      <fieldset>
  			<legend>{title}</legend>
        <div data-row-span="2">
          {this.showKey()}
          {this.showValue()}
        </div>
        <div data-row-span="2">
          {this.showEffectiveFrom()}
          {this.showEffectiveTo()}
        </div>
        <div data-row-span="1">
          {this.showRemarks()}
        </div>
      </fieldset>
      <br/>
      <Button
        type="submit"
        color="primary"
        disabled={!this.formValid}
        hidden={!authzn.allowsEdit}
        title={this.validationMsg}
        >Save</Button>
      <Button
        color="link"
        ><Link
          to="/durations"
          className="text-danger"
          title="Go to Durations"
          >Cancel</Link></Button>
    </Form>
  }

  handleSubmit(event) {
    const { model } = this.state
    event.preventDefault()
    this.setState({ submitted: true })
    this.props.saveChanges(model)
  }

  changedProps() {
    const { model } = this.state
    const { location } = this.props
    let modelDB = location.state.model
    let props = []

    // check for changes in model props
    for(const prop in model) {
      if( prop === 'id') continue // exclude 'id' from comparision
      if(modelDB[prop] !== model[prop]) { // if data is changed wrt data in database
        props.push(prop)
      }
    }
    return props;
  }

  showKey() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Key</Label>
        <Input
          id="key"
          type="text"
          name="key"
          value={model.key}
          placeholder="Key string here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.key
          && <FormText color="danger">Key is required</FormText>}
			</div>
  }

  handleChange(event) {
    const { name, value } = event.target
    const { model } = this.state
    this.setState({
        model: {
          ...model,
          [name]: value
        }
    })
  }

  showValue() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Value</Label>
        <Input
          id="value"
          type="text"
          name="value"
          value={model.value}
          placeholder="Value string here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.value
          && <FormText color="danger">Value is required</FormText>}
			</div>
  }

  showEffectiveFrom() {
    const { model } = this.state
    return <div data-field-span="1">
				<Label>Effective From Date</Label>
        <Input
          id="effectiveFrom"
          type="date"
          name="effective_from"
          value={model.effective_from}
          placeholder="Effective Start Date of Key and its Value"
          className="inputField"
          onChange={this.handleChange}
        />
			</div>
  }
  showEffectiveTo() {
    const { model } = this.state
    return <div data-field-span="1">
				<Label>Effective To Date</Label>
        <Input
          id="effectiveTo"
          type="date"
          name="effective_to"
          value={model.effective_to}
          placeholder="Effective End Date of Key and its Value"
          className="inputField"
          onChange={this.handleChange}
        />
			</div>
  }

  showRemarks() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Remarks</Label>
        <Input
          id="remarks"
          type="text"
          name="remarks"
          value={model.remarks}
          placeholder="Remarks string here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.remarks
          && <FormText color="danger">Remarks is required</FormText>}
			</div>
  }

}

function mapStateToProps(state) {
  const { alert, authorizations, users } = state
  const authzn = authorizations[module]
  return {
    alert,
    authzn,
    users
  }
}

function mapDispatchToProps(dispatch) {
  return {
    saveChanges: (model) => {
      dispatch(actions.saveChanges(model))
    },
    error: (msg) => {
      dispatch(alertActions.error(msg))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(DurationDetails)
