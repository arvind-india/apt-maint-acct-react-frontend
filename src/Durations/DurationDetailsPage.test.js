// DurationsPage.test.js0]

import React from 'react'
import { shallow } from 'enzyme'

import { DurationDetails } from './DurationDetailsPage'

const mockDurations = [
  {
    id: 1,
    key: 'guest',
    value: 'guest',
    effective_from: 'user',
    effective_to: 'guest',
    remarks: '2018-03-31'
  },
  {
    id: 2,
    key: 'og1',
    value: 'g1',
    effective_from: 'user',
    effective_to: 'owner',
    remarks: '2018-04-31'
  }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    saveChanges: jest.fn(),
    error: jest.fn(),
    durations: {
      items: []
    }
  },
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    saveChanges: jest.fn(),
    error: jest.fn(),
    durations: {
      items: []
    }
  }
]
const mockLocations = [
  {
    state: {
      model: mockDurations[0],
      submitted: false,
      adding: false
    }
  },
  {
    state: {
      model: mockDurations[1],
      submitted: false,
      adding: false
    }
  }
]

const mocks = [
  {
    key: {name: 'key', value: mockDurations[0].key},
    value: {name: 'value', value: mockDurations[0].value},
    effectiveFrom: {name: 'effective_from', value: mockDurations[0].effective_from},
    effectiveTo: {name: 'effective_to', value: mockDurations[0].effective_to},
    remarks: {name: 'remarks', value: mockDurations[0].remarks}
  },
  {
    key: {name: 'key', value: mockDurations[1].key},
    value: {name: 'value', value: mockDurations[1].value},
    effectiveFrom: {name: 'effective_from', value: mockDurations[1].effective_from},
    effectiveTo: {name: 'effective_to', value: mockDurations[1].effective_to},
    remarks: {name: 'remarks', value: mockDurations[1].remarks}
  }
]

const component = shallow(<DurationDetails {...mockProps[0]} location={mockLocations[0]}/>)

describe('DurationsPage params test', () => {
  it('should render DurationsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('model')).toBeDefined()
  })
  it('should have submitted state', () => {
    expect(component.state('submitted')).toBeDefined()
  })
  it('should have adding state', () => {
    expect(component.state('adding')).toBeDefined()
  })
})


describe('DurationDetailsPage events test 1', () => {

  it('should be called with required states as arguments', () => {
    component.find('#key').simulate('change', {target: mocks[0].key})
    component.find('#value').simulate('change', {target: mocks[0].value})
    component.find('#effectiveFrom').simulate('change', {target: mocks[0].effectiveFrom})
    component.find('#effectiveTo').simulate('change', {target: mocks[0].effectiveTo})
    component.find('#remarks').simulate('change', {target: mocks[0].remarks})
    component.find('#durationDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[0].saveChanges.mock.calls[0][0]).toEqual(mockDurations[0])
  })

})


const component2 = shallow(<DurationDetails {...mockProps[1]} location={mockLocations[1]}/>)
describe('DurationDetailsPage events test 2', () => {
  it('should be called with required states as arguments', () => {
    component2.find('#key').simulate('change', {target: mocks[1].key})
    component2.find('#value').simulate('change', {target: mocks[1].value})
    component2.find('#effectiveFrom').simulate('change', {target: mocks[1].effectiveFrom})
    component2.find('#effectiveTo').simulate('change', {target: mocks[1].effectiveTo})
    component2.find('#remarks').simulate('change', {target: mocks[1].remarks})
    component2.find('#durationDetailsForm').simulate('submit', {preventDefault() {}})
    expect(mockProps[1].saveChanges.mock.calls[0][0]).toEqual(mockDurations[1])
  })
})
