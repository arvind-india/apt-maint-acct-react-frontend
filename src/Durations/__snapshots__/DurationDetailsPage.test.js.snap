// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h2>
        Duration Details
    </h2>
    <FlashMessage
        color={undefined}
        delay={2100}
        text="testingAlert"
    />
    <Form
        className="grid-form"
        id="durationDetailsForm"
        onSubmit={[Function]}
        tag="form"
    >
        <fieldset>
            <legend>
                Edit
            </legend>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Key
                    </Label>
                    <Input
                        className="inputField"
                        id="key"
                        name="key"
                        onChange={[Function]}
                        placeholder="Key string here"
                        tag="p"
                        type="text"
                        value="guest"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Value
                    </Label>
                    <Input
                        className="inputField"
                        id="value"
                        name="value"
                        onChange={[Function]}
                        placeholder="Value string here"
                        tag="p"
                        type="text"
                        value="guest"
                    />
                </div>
            </div>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Effective From Date
                    </Label>
                    <Input
                        className="inputField"
                        id="effectiveFrom"
                        name="effective_from"
                        onChange={[Function]}
                        placeholder="Effective Start Date of Key and its Value"
                        tag="p"
                        type="date"
                        value="user"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Effective To Date
                    </Label>
                    <Input
                        className="inputField"
                        id="effectiveTo"
                        name="effective_to"
                        onChange={[Function]}
                        placeholder="Effective End Date of Key and its Value"
                        tag="p"
                        type="date"
                        value="guest"
                    />
                </div>
            </div>
            <div
                data-row-span="1"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Remarks
                    </Label>
                    <Input
                        className="inputField"
                        id="remarks"
                        name="remarks"
                        onChange={[Function]}
                        placeholder="Remarks string here"
                        tag="p"
                        type="text"
                        value="2018-03-31"
                    />
                </div>
            </div>
        </fieldset>
        <br />
        <Button
            color="primary"
            disabled={true}
            hidden={false}
            tag="button"
            title="No changes to save"
            type="submit"
        >
            Save
        </Button>
        <Button
            color="link"
            tag="button"
        >
            <Link
                className="text-danger"
                replace={false}
                title="Go to Durations"
                to="/durations"
            >
                Cancel
            </Link>
        </Button>
    </Form>
</div>,
]
`;
